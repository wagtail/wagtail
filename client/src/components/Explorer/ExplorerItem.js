import PropTypes from 'prop-types';
import React from 'react';

import { ADMIN_URLS, STRINGS } from '../../config/wagtailConfig';
import Icon from '../../components/Icon/Icon';
import Button from '../../components/Button/Button';
import PublicationStatus from '../../components/PublicationStatus/PublicationStatus';

// Hoist icons in the explorer item, as it is re-rendered many times.
const childrenIcon = (
  <Icon name="folder-inverse" className="icon--menuitem" />
);

/**
 * One menu item in the page explorer, with different available actions
 * and information depending on the metadata of the page.
 */
const ExplorerItem = ({ item, onClick }) => {
  const { id, admin_display_title: title, meta } = item;
  const hasChildren = meta.children.count > 0;
  const isPublished = meta.status.live && !meta.status.has_unpublished_changes;

  return (
    <div className="c-explorer__item">
      <Button href={`${ADMIN_URLS.PAGES}${id}/`} className="c-explorer__item__link">
        {hasChildren ? childrenIcon : null}

        <h3 className="c-explorer__item__title">
          {title}
        </h3>

        {!isPublished ? (
          <span className="c-explorer__meta">
            <PublicationStatus status={meta.status} />
          </span>
        ) : null}
      </Button>
      <Button
        href={`${ADMIN_URLS.PAGES}${id}/edit/`}
        className="c-explorer__item__action c-explorer__item__action--small"
      >
        <Icon name="edit" title={STRINGS.EDIT_PAGE.replace('{title}', title)} className="icon--item-action" />
      </Button>
      {hasChildren ? (
        <Button
          className="c-explorer__item__action"
          onClick={onClick}
          href={`${ADMIN_URLS.PAGES}${id}/`}
        >
          <Icon
            name="arrow-right"
            title={STRINGS.VIEW_CHILD_PAGES_OF_PAGE.replace('{title}', title)}
            className="icon--item-action"
          />
        </Button>
      ) : null}
    </div>
  );
};

ExplorerItem.propTypes = {
  item: PropTypes.shape({
    id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
    admin_display_title: PropTypes.string.isRequired,
    meta: PropTypes.shape({
      status: PropTypes.object.isRequired,
    }).isRequired,
  }).isRequired,
  onClick: PropTypes.func.isRequired,
};

export default ExplorerItem;
