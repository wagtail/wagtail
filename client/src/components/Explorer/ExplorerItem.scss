.c-explorer__item {
    display: flex;
    flex-flow: row nowrap;
    border-bottom: 1px solid $c-explorer-bg-dark;
}

.c-explorer__item__link {
    display: inline-flex;
    align-items: center;
    flex-grow: 1;
    padding: 1.45em 1em;
    cursor: pointer;

    &:focus {
        background: $c-explorer-bg-active;
        color: $color-white;
        outline: none;
    }

    // Overrides for default link hover.
    &:hover {
        color: $color-white;
    }

    @include hover {
        background: $c-explorer-bg-active;
    }

    @include medium {
        padding: 1.45em 1.75em;
    }
}

.c-explorer__item__link .icon {
    font-size: 2em;
    color: $c-explorer-secondary;
    margin-right: 0.5rem;
}

.c-explorer__item__title {
    margin: 0;
    color: $color-white;
    display: inline-block;
}

.c-explorer__item__action {
    display: inline-flex;
    align-items: center;
    justify-content: center;
    flex-shrink: 0;
    width: 50px;
    padding: 0 0.5em;
    line-height: 1;
    font-size: 2em;
    cursor: pointer;
    color: $c-explorer-secondary;
    border: 0;
    border-left: solid 1px $c-explorer-bg-dark;

    &:focus {
        background: $c-explorer-bg-active;
        color: $color-white;
        outline: none;
    }

    // Overrides for default link hover.
    &:hover {
        color: $c-explorer-secondary;
    }

    @include hover {
        background: $c-explorer-bg-active;
        color: $color-white;
    }

    .icon::before {
        margin-right: 0;
    }
}

.c-explorer__item__action--small {
    font-size: 1.2em;
}

.c-explorer__meta {
    margin-left: 0.5rem;
    color: $c-explorer-secondary;
    font-size: 12px;
}
