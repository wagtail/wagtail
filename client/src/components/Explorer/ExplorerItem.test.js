import React from 'react';
import { shallow } from 'enzyme';

import ExplorerItem from './ExplorerItem';

const mockProps = {
  item: {
    id: 5,
    admin_display_title: 'test',
    meta: {
      latest_revision_created_at: null,
      status: {
        live: true,
        status: 'test',
        has_unpublished_changes: false,
      },
      descendants: {
        count: 0,
      },
      children: {
        count: 0,
      }
    },
  },
  onClick: () => {},
};

describe('ExplorerItem', () => {
  it('exists', () => {
    expect(ExplorerItem).toBeDefined();
  });

  it('renders', () => {
    expect(shallow(<ExplorerItem {...mockProps} />)).toMatchSnapshot();
  });

  it('children', () => {
    const props = Object.assign({}, mockProps);
    props.item.meta.children.count = 5;
    expect(shallow(<ExplorerItem {...props} />)).toMatchSnapshot();
  });

  it('should show a publication status with unpublished changes', () => {
    const props = Object.assign({}, mockProps);
    props.item.meta.status.has_unpublished_changes = true;
    expect(shallow(<ExplorerItem {...props} />)).toMatchSnapshot();
  });

  it('should show a publication status if not live', () => {
    const props = Object.assign({}, mockProps);
    props.item.meta.status.live = false;
    expect(shallow(<ExplorerItem {...props} />)).toMatchSnapshot();
  });
});
