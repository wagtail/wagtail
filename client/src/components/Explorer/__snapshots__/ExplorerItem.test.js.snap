// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExplorerItem children 1`] = `
<div
  className="c-explorer__item"
>
  <Button
    className="c-explorer__item__link"
    href="/admin/pages/5/"
  >
    <Icon
      className="icon--menuitem"
      name="folder-inverse"
    />
    <h3
      className="c-explorer__item__title"
    >
      test
    </h3>
  </Button>
  <Button
    className="c-explorer__item__action c-explorer__item__action--small"
    href="/admin/pages/5/edit/"
  >
    <Icon
      className="icon--item-action"
      name="edit"
      title="Edit 'test'"
    />
  </Button>
  <Button
    className="c-explorer__item__action"
    href="/admin/pages/5/"
    onClick={[Function]}
  >
    <Icon
      className="icon--item-action"
      name="arrow-right"
      title="View child pages of 'test'"
    />
  </Button>
</div>
`;

exports[`ExplorerItem renders 1`] = `
<div
  className="c-explorer__item"
>
  <Button
    className="c-explorer__item__link"
    href="/admin/pages/5/"
  >
    <h3
      className="c-explorer__item__title"
    >
      test
    </h3>
  </Button>
  <Button
    className="c-explorer__item__action c-explorer__item__action--small"
    href="/admin/pages/5/edit/"
  >
    <Icon
      className="icon--item-action"
      name="edit"
      title="Edit 'test'"
    />
  </Button>
</div>
`;

exports[`ExplorerItem should show a publication status if not live 1`] = `
<div
  className="c-explorer__item"
>
  <Button
    className="c-explorer__item__link"
    href="/admin/pages/5/"
  >
    <Icon
      className="icon--menuitem"
      name="folder-inverse"
    />
    <h3
      className="c-explorer__item__title"
    >
      test
    </h3>
    <span
      className="c-explorer__meta"
    >
      <PublicationStatus
        status={
          Object {
            "has_unpublished_changes": true,
            "live": false,
            "status": "test",
          }
        }
      />
    </span>
  </Button>
  <Button
    className="c-explorer__item__action c-explorer__item__action--small"
    href="/admin/pages/5/edit/"
  >
    <Icon
      className="icon--item-action"
      name="edit"
      title="Edit 'test'"
    />
  </Button>
  <Button
    className="c-explorer__item__action"
    href="/admin/pages/5/"
    onClick={[Function]}
  >
    <Icon
      className="icon--item-action"
      name="arrow-right"
      title="View child pages of 'test'"
    />
  </Button>
</div>
`;

exports[`ExplorerItem should show a publication status with unpublished changes 1`] = `
<div
  className="c-explorer__item"
>
  <Button
    className="c-explorer__item__link"
    href="/admin/pages/5/"
  >
    <Icon
      className="icon--menuitem"
      name="folder-inverse"
    />
    <h3
      className="c-explorer__item__title"
    >
      test
    </h3>
    <span
      className="c-explorer__meta"
    >
      <PublicationStatus
        status={
          Object {
            "has_unpublished_changes": true,
            "live": true,
            "status": "test",
          }
        }
      />
    </span>
  </Button>
  <Button
    className="c-explorer__item__action c-explorer__item__action--small"
    href="/admin/pages/5/edit/"
  >
    <Icon
      className="icon--item-action"
      name="edit"
      title="Edit 'test'"
    />
  </Button>
  <Button
    className="c-explorer__item__action"
    href="/admin/pages/5/"
    onClick={[Function]}
  >
    <Icon
      className="icon--item-action"
      name="arrow-right"
      title="View child pages of 'test'"
    />
  </Button>
</div>
`;
